providers/pvrdma/cq.c

```diff
@@ -0,0 +1,287 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <infiniband/arch.h>
+
+#include "pvrdma.h"
+
+enum {
+	CQ_OK =  0,
+	CQ_EMPTY = -1,
+	CQ_POLL_ERR = -2,
+};
+
+enum {
+	PVRDMA_CQE_IS_SEND_MASK = 0x40,
+	PVRDMA_CQE_OPCODE_MASK = 0x1f,
+};
+
+int pvrdma_alloc_cq_buf(struct pvrdma_device *dev, struct pvrdma_cq *cq,
+			struct pvrdma_buf *buf, int entries)
+{
+	if (pvrdma_alloc_buf(buf, cq->offset +
+			     entries * (sizeof(struct pvrdma_cqe)),
+			     dev->page_size))
+		return -1;
+	memset(buf->buf, 0, buf->length);
+
+	return 0;
+}
+
+static struct pvrdma_cqe *get_cqe(struct pvrdma_cq *cq, int entry)
+{
+	return cq->buf.buf + cq->offset +
+	       entry * (sizeof(struct pvrdma_cqe));
+}
+
+static int pvrdma_poll_one(struct pvrdma_cq *cq,
+			   struct pvrdma_qp **cur_qp,
+			   struct ibv_wc *wc)
+{
+	struct pvrdma_context *ctx = to_vctx(cq->ibv_cq.context);
+	int has_data;
+	unsigned int head;
+	int tried = 0;
+	struct pvrdma_cqe *cqe;
+
+retry:
+	has_data = pvrdma_idx_ring_has_data(&cq->ring_state->rx,
+					    cq->cqe_cnt, &head);
+	if (has_data == 0) {
+		unsigned int val;
+
+		if (tried)
+			return CQ_EMPTY;
+
+		/* Pass down POLL to give physical HCA a chance to poll. */
+		val = cq->cqn | PVRDMA_UAR_CQ_POLL;
+		pvrdma_write_uar_cq(ctx->uar, val);
+
+		tried = 1;
+		goto retry;
+	} else if (has_data == -1) {
+		return CQ_POLL_ERR;
+	}
+
+	cqe = get_cqe(cq, head);
+	if (!cqe)
+		return CQ_EMPTY;
+
+	rmb();
+
+	if (ctx->qp_tbl[cqe->qp & 0xFFFF])
+		*cur_qp = (struct pvrdma_qp *)ctx->qp_tbl[cqe->qp & 0xFFFF];
+	 else
+		return CQ_POLL_ERR;
+
+	wc->opcode = pvrdma_wc_opcode_to_ibv(cqe->opcode);
+	wc->status = pvrdma_wc_status_to_ibv(cqe->status);
+	wc->wr_id = cqe->wr_id;
+	wc->qp_num = (*cur_qp)->ibv_qp.qp_num;
+	wc->byte_len = cqe->byte_len;
+	wc->imm_data = cqe->imm_data;
+	wc->src_qp = cqe->src_qp;
+	wc->wc_flags = cqe->wc_flags;
+	wc->pkey_index = cqe->pkey_index;
+	wc->slid = cqe->slid;
+	wc->sl = cqe->sl;
+	wc->dlid_path_bits = cqe->dlid_path_bits;
+	wc->vendor_err = 0;
+
+	/* Update shared ring state. */
+	pvrdma_idx_ring_inc(&(cq->ring_state->rx.cons_head), cq->cqe_cnt);
+
+	return CQ_OK;
+}
+
+int pvrdma_poll_cq(struct ibv_cq *ibcq, int num_entries, struct ibv_wc *wc)
+{
+	struct pvrdma_cq *cq = to_vcq(ibcq);
+	struct pvrdma_qp *qp;
+	int npolled = 0;
+
+	if (num_entries < 1 || wc == NULL)
+		return 0;
+
+	pthread_spin_lock(&cq->lock);
+
+	for (npolled = 0; npolled < num_entries; ++npolled) {
+		if (pvrdma_poll_one(cq, &qp, wc + npolled) != CQ_OK)
+			break;
+	}
+
+	pthread_spin_unlock(&cq->lock);
+
+	return npolled;
+}
+
+void __pvrdma_cq_clean(struct pvrdma_cq *cq, uint32_t qpn)
```

Comment:
Thanks for talking a look! I'll drop the __ and rename this since we already have a pvrdma_cq_clean function below which uses this.